cmd/thanos/compact.go

```diff
@@ -223,6 +235,60 @@ func runCompact(
 		}, func(error) {
 			cancel()
 		})
+
+		g.Add(func() error {
+			if retention.Seconds() == 0 {
+				level.Info(logger).Log("msg", "retention is disabled")
+
+				<-ctx.Done()
+			}
+
+			retentionFunc := func() error {
+				//Wait for compaction job to finish
+				mutex.Lock()
+				defer mutex.Unlock()
+
+				level.Info(logger).Log("msg", "start retention", "retention", retention)
+				err := bkt.Iter(ctx, "", func(name string) error {
+					id, ok := block.IsBlockDir(name)
+					if !ok {
+						return nil
+					}
+					m, err := block.DownloadMeta(ctx, logger, bkt, id)
+					if err != nil {
+						return errors.Wrap(err, "download metadata")
+					}
+
+					maxTime := time.Unix(m.MaxTime/1000, 0)
+					if time.Now().After(maxTime.Add(retention)) {
+						level.Info(logger).Log("msg", "deleting block", "id", id, "maxTime", maxTime.String())
```

Comment:
This is really really critical path. 

We DON'T want to remove metric data by accident, so tests are **must have** here.

Reply:
Worth to log minTime as well